src/vision/status_quo/barbara_wants_single_threaded_optimizations.md

```diff
@@ -0,0 +1,126 @@
+
+### Rejected: both single- and multi-threaded versions
+
+Barbara considers her options to make the crate usable by the multi-threaded system service. She decides against making `IpcHandle` available in both single-threaded and multi-threaded versions. To do this generically would require a lot of boilerplate. For example, it would require manually duplicating APIs which would need to have a `Send` bound in the multi-threaded case:
+
+```rust
+impl LocalIpcHandle {
+    fn spawn_on_reply<F: Future + 'static>(&mut self, to_spawn: impl Fn(IpcReply) -> F) {
+        // ...
+    }
+}
+
+impl SendIpcHandle {
+    fn spawn_on_reply<F: Future + Send + 'static>(&mut self, to_spawn: impl Fn(IpcReply) -> F) {
+        // ...
+    }
+}
+```
+
+### Accepted: only implement multi-threaded version
+
+Barbara decides it's not worth the effort to duplicate so much of the crate's functionality, and decides to make the whole library thread-safe:
+
+```rust
+pub struct IpcHandle {
+    cache_storage: Arc<Mutex<IpcCache>>,
+    // ...
+}
+
+struct IpcCache { /* ... */ }
+```
+
+This requires her to migrate her original system service to use multi-threaded types when interacting with the library. Before the change her service uses only single-threaded reference counting and interior mutability:
+
+```rust
+#[derive(Clone)]
+struct ClientBroker {
+    state: Rc<RefCell<ClientState>>,
+}
+
+impl ClientBroker {
+    fn start_serving_clients(self) {
+        let mut ipc_handle = self.make_ipc_handle_for_new_clients();
+        ipc_handle.spawn_on_reply(move |reply| shared_state.clone().serve_client(reply));
+        LocalExecutor::new().run_singlethreaded(ipc_handle.listen());
+    }
+
+    fn make_ipc_handle_for_new_clients(&self) { /* ... */ }
+    async fn serve_client(self, reply: IpcReply) { /* accesses interior mutability... */ }
+}
+```
+
+In order to be compatible with her own crate, Barbara needs to wrap the shared state of her service behind multi-threaded reference counting and synchronization:
+
+```rust
+#[derive(Clone)]
+struct ClientBroker {
+    state: Arc<Mutex<ClientState>>,
+}
+
+impl ClientBroker { /* nothing changed */ }
+```
+
+This incurs some performance overhead when cloning the `Arc` and when accessing the `Mutex`. The former is cheap when uncontended on x86 but will have different performance characteristics on e.g. ARM platforms. The latter's overhead varies depending on the kind of `Mutex` used, e.g. an uncontended `parking_lot::Mutex` may only need a few atomic instructions to acquire it. Acquiring many platforms' `std::sync::Mutex` is much more expensive than a few atomics. This overhead is usually not very high, but it does pollute shared resources like cache lines and is multiplied by the number of single-threaded services which make use of such a library.
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
```

Comment:
To really drive this point home, maybe finish with that the status quo is using thread-safe data structures in the vast majority of cases, since the added maintenance of dual implementations isn't feasible? Paying for things you don't use (memory barriers, atomics and kernel locks) is arguably a violation of the zero-cost abstraction principle.